File: src/main/scala/cn/cjie/scala/akka/mapreduce/MapReduceApplication.scala
package cn.cjie.scala.akka.mapreduce

import akka.actor.{Props, ActorSystem}
import akka.util.Timeout
import scala.concurrent.Await
import scala.collection.mutable.ArrayBuffer
import java.util.concurrent.TimeUnit
import akka.pattern.ask

/**
 * Created with IntelliJ IDEA.
 * User: hucj
 * Date: 14-10-27
 * Time: 下午5:21
 * To change this template use File | Settings | File Templates.
 */

sealed trait MapReduceMessage
case class WordCount(word: String, count: Int) extends
MapReduceMessage
case class MapData(dataList: ArrayBuffer[WordCount]) extends
MapReduceMessage
case class ReduceData(reduceDataMap: Map[String, Int]) extends
MapReduceMessage
case class Result extends MapReduceMessage


object MapReduceApplication  extends  App {

  val _system = ActorSystem("MapReduceApp")
  val master = _system.actorOf(Props[MasterActor], name = "master")
  implicit val timeout = Timeout(5, TimeUnit.SECONDS)
  master ! "The quick brown fox tried to jump over the lazy dog and fell on the dog"
  master ! "Dog is man's best friend"
  master ! "Dog and Fox belong to the same family"

  Thread.sleep(500)
  val future = (master ? Result).mapTo[String]
  val result = Await.result(future, timeout.duration)
  println(result)
  _system.shutdown

}

File: src/main/java/com/cjie/akka/mapreduce/MapReduceApplication.java
package com.cjie.akka.mapreduce;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.pattern.Patterns;
import akka.util.Timeout;
import scala.concurrent.Await;
import scala.concurrent.Future;

import java.util.concurrent.TimeUnit;

/**
 * Created with IntelliJ IDEA.
 * User: hucj
 * Date: 14-10-27
 * Time: 下午4:52
 * To change this template use File | Settings | File Templates.
 */
public class MapReduceApplication {

    public static void main(String[] args) throws Exception {
        Timeout timeout = new Timeout(5, TimeUnit.SECONDS);
        ActorSystem _system = ActorSystem.create("MapReduceApp");
        ActorRef master = _system.actorOf(new
                Props(MasterActor.class),
                "master");
        master.tell("The quick brown fox tried to jump over the lazy dog and fell on the dog", master);
        master.tell("Dog is man's best friend", master);
        master.tell("Dog and Fox belong to the same family", master);
        Thread.sleep(5000);
        Future<Object> future = Patterns.ask(master, new
                Result(), timeout);
        String result = (String) Await.result(future,
                timeout.duration());
        System.out.println(result);
        _system.shutdown();
    }
}
